File: packages/node/test/contract.test.ts
/// <reference types="vitest" />
import { describe, expect, it } from "vitest";
import {
  createKit,
  Provider,
  GenerateInput,
} from "../src/index.js";
import type { FetchLike } from "../src/core/types.js";

const STANDARD_TEXT = "Unified response";

const baseMessages: GenerateInput["messages"] = [
  {
    role: "system",
    content: [{ type: "text", text: "You are a helpful assistant." }],
  },
  {
    role: "user",
    content: [{ type: "text", text: "Say hello." }],
  },
];

const fetchImpl: FetchLike = async (input, init) => {
  const url = typeof input === "string" ? input : input.toString();
  if (url.includes("api.openai.com")) {
    return jsonResponse(buildChatCompletion(STANDARD_TEXT));
  }
  if (url.includes("api.anthropic.com")) {
    return jsonResponse(buildAnthropicResponse(STANDARD_TEXT));
  }
  if (url.includes("api.x.ai")) {
    return jsonResponse(buildChatCompletion(STANDARD_TEXT));
  }
  if (url.includes("generativelanguage.googleapis.com")) {
    if (url.includes(":generateContent")) {
      return jsonResponse(buildGeminiResponse(STANDARD_TEXT));
    }
  }
  throw new Error(`No mock response for ${url}`);
};

const kit = createKit({
  providers: {
    [Provider.OpenAI]: { apiKey: "test-openai-key" },
    [Provider.Anthropic]: { apiKey: "test-anthropic-key" },
    [Provider.XAI]: { apiKey: "test-xai-key" },
    [Provider.Google]: { apiKey: "test-google-key" },
  },
  httpClient: fetchImpl,
});

describe("contract generate normalization", () => {
  const input: Omit<GenerateInput, "provider"> = {
    model: "test-model",
    messages: baseMessages,
  };

  for (const provider of [
    Provider.OpenAI,
    Provider.Anthropic,
    Provider.Google,
    Provider.XAI,
  ]) {
    it(`normalizes output for ${provider}`, async () => {
      const output = await kit.generate({ ...input, provider });
      expect(output.text).toBe(STANDARD_TEXT);
      expect(output.finishReason).toBeDefined();
    });
  }
});

function jsonResponse(body: unknown): Response {
  return new Response(JSON.stringify(body), {
    status: 200,
    headers: { "content-type": "application/json" },
  });
}

function buildChatCompletion(text: string) {
  return {
    id: "chatcmpl-1",
    choices: [
      {
        finish_reason: "stop",
        message: {
          content: [{ type: "text", text }],
          tool_calls: [],
        },
      },
    ],
    usage: {
      prompt_tokens: 5,
      completion_tokens: 10,
      total_tokens: 15,
    },
  };
}

function buildAnthropicResponse(text: string) {
  return {
    id: "msg_1",
    content: [{ type: "text", text }],
    stop_reason: "end_turn",
    usage: { input_tokens: 5, output_tokens: 10 },
  };
}

function buildGeminiResponse(text: string) {
  return {
    candidates: [
      {
        content: {
          parts: [{ text }],
        },
        finishReason: "STOP",
      },
    ],
    usageMetadata: {
      promptTokenCount: 1,
      candidatesTokenCount: 1,
      totalTokenCount: 2,
    },
  };
}

File: packages/go/contract_test.go
package aikit

import (
	"bytes"
	"context"
	"io"
	"net/http"
	"strings"
	"testing"
)

type roundTripFunc func(*http.Request) (*http.Response, error)

func (f roundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
	return f(req)
}

func TestHubGenerateContract(t *testing.T) {
	client := &http.Client{
		Transport: roundTripFunc(func(req *http.Request) (*http.Response, error) {
			switch {
			case strings.Contains(req.URL.Host, "api.openai.com") && strings.Contains(req.URL.Path, "/responses"):
				return jsonHTTPResponse(`{"status":"completed","output":[{"content":[{"type":"output_text","text":"Unified response"}]}],"usage":{"input_tokens":1,"output_tokens":1,"total_tokens":2}}`), nil
			case strings.Contains(req.URL.Host, "api.openai.com"):
				return jsonHTTPResponse(`{"choices":[{"finish_reason":"stop","message":{"content":[{"type":"text","text":"Unified response"}],"tool_calls":[]}}],"usage":{"prompt_tokens":1,"completion_tokens":1,"total_tokens":2}}`), nil
			case strings.Contains(req.URL.Host, "api.anthropic.com"):
				return jsonHTTPResponse(`{"content":[{"type":"text","text":"Unified response"}],"stop_reason":"end_turn","usage":{"input_tokens":1,"output_tokens":1}}`), nil
			case strings.Contains(req.URL.Host, "api.x.ai"):
				return jsonHTTPResponse(`{"choices":[{"finish_reason":"stop","message":{"content":[{"type":"text","text":"Unified response"}],"tool_calls":[]}}]}`), nil
			case strings.Contains(req.URL.Host, "generativelanguage.googleapis.com"):
				return jsonHTTPResponse(`{"candidates":[{"content":{"parts":[{"text":"Unified response"}]},"finishReason":"STOP"}],"usageMetadata":{"promptTokenCount":1,"candidatesTokenCount":1,"totalTokenCount":2}}`), nil
			default:
				t.Fatalf("unexpected request: %s", req.URL.String())
				return nil, nil
			}
		}),
	}

	hub, err := New(Config{
		OpenAI:     &OpenAIConfig{APIKey: "openai", DefaultUseResponses: true},
		Anthropic:  &AnthropicConfig{APIKey: "anthropic"},
		XAI:        &XAIConfig{APIKey: "xai", CompatibilityMode: "openai"},
		Google:     &GoogleConfig{APIKey: "google"},
		HTTPClient: client,
	})
	if err != nil {
		t.Fatalf("new hub: %v", err)
	}

	baseInput := GenerateInput{
		Model:    "test-model",
		Messages: []Message{{Role: "user", Content: []ContentPart{{Type: "text", Text: "hello"}}}},
	}
	providers := []Provider{ProviderOpenAI, ProviderAnthropic, ProviderGoogle, ProviderXAI}
	for _, provider := range providers {
		input := baseInput
		input.Provider = provider
		if provider == ProviderOpenAI {
			input.ResponseFormat = &ResponseFormat{
				Type: "json_schema",
				JsonSchema: &JsonSchemaFormat{
					Name:   "demo",
					Schema: map[string]interface{}{"type": "object"},
					Strict: true,
				},
			}
		}
		output, err := hub.Generate(context.Background(), input)
		if err != nil {
			t.Fatalf("generate %s: %v", provider, err)
		}
		if output.Text != "Unified response" {
			t.Fatalf("provider %s expected normalized text, got %s", provider, output.Text)
		}
	}
}

func jsonHTTPResponse(body string) *http.Response {
	return &http.Response{
		StatusCode: 200,
		Header:     http.Header{"Content-Type": []string{"application/json"}},
		Body:       io.NopCloser(bytes.NewBufferString(body)),
	}
}
